CPP20.md

```diff
@@ -0,0 +1,274 @@
+which is a sequence of logical operations and operands that specifies requirements on template arguments.
+
+Named sets of such requirements are called concepts. Each concept is a predicate, evaluated at compile time,
+and becomes a part of the interface of a template where it is used as a constraint.
+
+```c++
+#include <string>
+#include <cstddef>
+using namespace std::literals;
+
+// Declaration of the concept "Hashable", which is satisfied by
+// any type T such that for values a of type T,
+// the expression std::hash<T>{}(a) compiles and its result is convertible to std::size_t
+template <typename T>
+concept Hashable = requires(T a) {
+    { std::hash<T>{}(a) } -> std::size_t;
+};
+
+struct meow {};
+
+template <Hashable T>
+void f(T); // constrained C++20 function template
+
+// Alternative ways to apply the same constraint:
+// template <typename T>
+// requires Hashable<T>
+// void f(T);
+//
+// template <typename T>
+// void f(T) requires Hashable<T>;
+
+int main() {
+  f("abc"s); // OK, std::string satisfies Hashable
+  f(meow{}); // Error: meow does not satisfy Hashable
+}
+```
+
+The intent of concepts is to model semantic categories (Number, Range, RegularFunction) rather than syntactic restrictions (HasPlus, Array).
+The definition of a concept must appear at namespace scope and has the form:
+`
+template <template-parameter-list>
+concept concept-name = constraint-expression;
+`
+```c++
+template <typename T, typename U>
+concept Derived = std::is_base_of<U, T>::value;
+template <typename T>
+concept Integral = std::is_integral<T>::value;
+template <typename T>
+concept SignedIntegral = Integral<T> && std::is_signed<T>::value;
+template <typename T>
+concept UnsignedIntegral = Integral<T> && !SignedIntegral<T>;
+
+template <typename T>
+using Ref = T&;
+
+template <typename T>
+concept C = requires {
+    typename T::inner; // required nested member name
+    typename S<T>;     // required class template specialization
+    typename Ref<T>;   // required alias template substitution
+};
+
+template <typename T, typename U>
+using CommonType = std::common_type_t<T, U>;
+
+template <typename T, typename U>
+concept Common = requires (T t, U u) {
+    typename CommonType<T, U>; // CommonType<T, U> is valid and names a type
+    { CommonType<T, U>{std::forward<T>(t)} };
+    { CommonType<T, U>{std::forward<U>(u)} };
+};
+
+template <typename T>
+concept Semiregular = DefaultConstructible<T> &&  CopyConstructible<T> && 
+    Destructible<T> && CopyAssignable<T> && requires(T a, size_t n) {
+    requires Same<T*, decltype(&a)>;  // nested: "Same<...> evaluates to true"
+    { a.~T() } noexcept;  // compound: "a.~T()" is a valid expression that doesn't throw
+    requires Same<T*, decltype(new T)>; // nested: "Same<...> evaluates to true"
+    requires Same<T*, decltype(new T[n])>; // nested
+    { delete new T };  // compound
+    { delete new T[n] }; // compound
+};
+```
```

Comment:
This section needs to be reworked to describe the syntax in more detail. I have a draft of this section in my local copy. Probably best to remove this code snippet for now.